File: trunk/osm/astar.pl
#! /usr/bin/perl -w

use strict;
use OSM::Map;
use Getopt::Long;
use Data::Dumper;
use Storable;

sub reconstruct_path {
    my $came_from = shift;
    my $goes_to = shift;
    my $current_node = shift;
    
    my @path = ();
    my @rev;
    my $cn = $current_node;
    
    while (defined($came_from->{$cn})) {
	$cn = $came_from->{$cn};
        push @rev,$cn;
    }
    while ($cn=pop(@rev)) {
	push @path,$cn;
    }
    push @path,$current_node;
    my $x = $current_node;
    if (defined($goes_to)) {
	while (exists($$goes_to{$x})) {
	    push @path,$$goes_to{$x};
	    $x=$$goes_to{$x};
	}
    }
    return  @path;
}

sub Astar {
    my $map = shift;
    my $startlat = shift;
    my $startlon = shift;
    my $goallat  = shift;
    my $goallon  = shift;
    my $vehicle = shift;
    
    my (%closedset,%goalset,%startset,%gs_score,%hs_score,%fs_score,%ds_score,%gg_score,%hg_score,%fg_score,%dg_score,%came_from,%goes_to);
    my (@openset,@f,@g,@h,@d,@to,@gl);
    
    $map->initRoute($vehicle);
    
    $map->fetchCoor($startlat,$startlon,1) unless $map->inboundCoor($startlat,$startlon);
    $map->fetchCoor($goallat,$goallon,1) unless $map->inboundCoor($goallat,$goallon);
    my $start = $map->findNode($startlat,$startlon,10);
    my $goal = $map->findNode($goallat,$goallon,10);
    print "start = $start, goal = $goal ".(defined($vehicle)?$vehicle:"")."\n";

    $startset{$start} = 1;
    $goalset{$goal} = 2;
    
    $gg_score{$goal} = $gs_score{$start} = 0;
    $fg_score{$goal} = $hg_score{$goal} = $fs_score{$start} = $hs_score{$start} = $map->calc_h_score($start,$goal,$vehicle);
    $dg_score{$goal} = $ds_score{$start} = $map->distance($start,$goal);    
    
    $openset[1] = \%startset;
    $openset[2] = \%goalset;
    $f[1] = \%fs_score;
    $f[2] = \%fg_score;
    $g[1] = \%gs_score;
    $g[2] = \%gg_score;
    $h[1] = \%hs_score;
    $h[2] = \%hg_score;
    $d[1] = \%ds_score;
    $d[2] = \%dg_score;
    $to[1] = \%came_from;
    $to[2] = \%goes_to;
    $gl[1] = $goal;
    $gl[2] = $start;
    
   while (keys(%startset) != 0 or keys(%goalset) != 0) {
	for (my $i=1;$i<=2;$i++) {
	    my $goal = $gl[$i];
	    my $xs = undef;
	    
	    for my $k (keys(%{$openset[$i]})) {
		$xs=$k unless defined($xs);
		$xs=$k if (${$f[$i]}{$k} < ${$f[$i]}{$xs});
	    }
	    if (defined($xs)) {
		if (exists($closedset{$xs}) and $closedset{$xs} != $i) {
		    my @p = reconstruct_path($to[1],$to[2],$xs);
		    return @p;
		}
		delete(${$openset[$i]}{$xs});
		$closedset{$xs} = $i;
#	        print "close $xs ".$f_score{$xs}."\n";                
		for my $y (@{$map->neighbours($xs)}) {
                    my $xs1 = $xs;
                    my $prevnode = ${$to[$i]}{$xs1};
		    $map->fetchNode($y) unless $map->inboundNode($y);
		    next if (defined($closedset{$y}));
                    my @ynb = @{$map->neighbours($y)};
                    while ($#ynb == 1) {
#                        print "Proceeding on $y from $xs1".(defined($prevnode) ?" and $prevnode":"")."\n";
                        ${$g[$i]}{$y} = ${$g[$i]}{$xs1} + ($i==1?$map->cost($xs1,$y,$prevnode): $map->cost($y,$xs1,$prevnode));
                        if ($ynb[0] == $xs1) {
                            ${$to[$i]}{$y} = $xs1;
                            $prevnode=$xs1;
                            $xs1=$y;
                            $y = $ynb[1];
                        } else {
                            ${$to[$i]}{$y} = $xs1;
                            $prevnode=$xs1;
                            $xs1=$y;
                            $y = $ynb[0];
                        }
                        $closedset{$xs1}=$i;
  		        $map->fetchNode($y) unless $map->inboundNode($y);
		        last if (defined($closedset{$y}));
                        @ynb = @{$map->neighbours($y)};
                    }
                    next if (defined($closedset{$y}));
		    my $tentative_g_score = ${$g[$i]}{$xs1} + ($i==1?$map->cost($xs1,$y,$prevnode):$map->cost($y,$xs1,$prevnode));
		    my $tentative_is_better;
		    unless (defined(${$openset[$i]}{$y})) {
			${$openset[$i]}{$y} = 1 ;
			$tentative_is_better = 1;
		    } elsif ($tentative_g_score < ${$g[$i]}{$y}) {
			$tentative_is_better = 1;
		    } else {
			$tentative_is_better = 0;
		    }
		    if ($tentative_is_better) {
			${$to[$i]}{$y} = $xs1;
			${$g[$i]}{$y} = $tentative_g_score;
			${$h[$i]}{$y} = $map->calc_h_score($y,$goal);
			${$d[$i]}{$y} = $map->distance($y,$goal);
			${$f[$i]}{$y} = ${$g[$i]}{$y}+${$h[$i]}{$y};
#			my $w = $map->getway($xs,$y);
#			print "$i $xs $y ",${$g[$i]}{$y}," ",${$h[$i]}{$y}," ",${$f[$i]}{$y},"\n" if $i==1;
#			print "$i $y $xs ",${$g[$i]}{$y}," ",${$h[$i]}{$y}," ",${$f[$i]}{$y},"\n" if $i==2;
		    }
		}
	    }
	    print "\r$xs ",${$d[$i]}{$xs};
#            print "$i xs=$xs openset=",join(", ",keys(%{$openset[$i]})),"\n";
#	     dump_openset($map,$openset[$i]);
	}
    }
    
#     return failure
    open FAIL,">astar.err.log";
    print FAIL "----------------------------\nOpenset\n----------------------------\n";
    print FAIL Dumper \@openset;
    print FAIL "----------------------------\nClosedset\n----------------------------\n";
    print FAIL Dumper \%closedset;
    print FAIL "----------------------------\nf\n----------------------------\n";
    print FAIL Dumper \@f;
    print FAIL "----------------------------\ng\n----------------------------\n";
    print FAIL Dumper \@g;
    print FAIL "----------------------------\nh\n----------------------------\n";
    print FAIL Dumper \@h;
    print FAIL "----------------------------\nto\n----------------------------\n";
    print FAIL Dumper \@to;
    print FAIL "----------------------------\ngl\n----------------------------\n";
    print FAIL Dumper \@gl;
    close FAIL;
    die "foutje";
}

my $filnum=0;
sub dump_openset {
	my $map = shift;
	my $openset = shift;
	
	my $f=$filnum++;
	$filnum %=10;
	if (open UIT,">astardmp-$filnum.csv") {
	        foreach my $o (keys(%{$openset})) {
	            printf UIT "%9.5f\t%9.5f\t%d\n",$map->getCoor($o),$o;
	    }
	    close UIT;
	}
}

sub print_path {
    my $map = shift;
    
    open PP,">>route.log";
    print PP "Route log:\n\n";
    
    my @p1=@_;
    print join(", ",@p1),"\n";
    print PP join(", ",@p1),"\n";
#    print " ",$map->node($p1[0])->{length} if defined($map->node($p1[0])->{length});

    my $oldp;
    my $oldw;
    my $dir;
    my $olddir;
    my $totdist=0;
    for my $p (@p1) {
        my $w;

	print "$p -> ";
	print PP "$p -> ";
	my @ws = $map->getways($p);
	if ($#ws == 0) {
	    $w = $ws[0];
	} else {
	    if (defined($oldp)) {
	        $w=$map->getway($oldp,$p);
	    }
	}
	if (defined($w)) {
	    print $w,": ";
	    print PP $w,": ";
            my $ww=$map->loadway($w);
	    my $mwt = $ww->{tag};
	    print $$mwt{name}," " if defined($$mwt{name});
	    print $$mwt{highway}," " if defined($$mwt{highway});
	    print $$mwt{maxspeed}," " if defined($$mwt{maxspeed});
	    print $$mwt{ref}," " if defined($$mwt{ref});
	    print PP $$mwt{name}," " if defined($$mwt{name});
	    print PP $$mwt{highway}," " if defined($$mwt{highway});
	    print PP $$mwt{maxspeed}," " if defined($$mwt{maxspeed});
	    print PP $$mwt{ref}," " if defined($$mwt{ref});
            if (defined($oldp)) {
                my $d = $map->dist($oldp,$p);
	        if (defined($d)) {
                    $olddir = $dir;
                    $dir = $map->direction($oldp,$p);
		    print "$dir $d ";
		    print PP "$dir $d ";
		    $totdist += $d;
	        }
	        print $totdist," ";
	        print PP $totdist," ";
            }
            $oldw=$w;
            my $locstr = $map->findLocation($p);
            print $locstr;
            print PP $locstr;
	}
	$oldp=$p;
	print "\n";
	print PP "\n";
    }
    print PP "\n===============End of log ==================\n\n";
    close PP;
}

my $new=0;
my $dbonly=0;
my $netonly=0;
my $confile='astarconf.xml';
my $mapfile = undef;
my $bbox = undef;
my $directory;

my %options = ('new' => \$new,
               'dbonly' => \$dbonly,
               'net' => \$netonly,
               'conf=s' => \$confile,
               'map=s'=>\$mapfile,
               'dir=s'=>\$directory,
               'bbox=s'=>\$bbox);
               
my $result = GetOptions(%options);

my $map = OSM::Map->new();

my @files = ();
if (defined($mapfile)) {
    $map->importOSMfile($mapfile);
} elsif (defined ($bbox)) {
    $map->importBbox($bbox);
} elsif (defined($directory)) {
    opendir(my $dh, $directory) || die "can't opendir $directory: $!";
    @files = grep { /map_.*\.osm$/ && -f "$directory/$_" } readdir($dh);
    map {s:^:$directory/: => $_} @files;
    closedir $dh;
    foreach my $f (@files) {
        $map->importOSMfile("$f");
    }
}

unless ($dbonly) {
###huis school
#print_path($map,Astar($map,52.297277,4.862030,52.29334,4.85876));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2935821,4.8593675,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2933,4.8588,'car'));

###Huis hockey
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2886,4.8508,'car'));

###Huis atletiek
#print_path($map,Astar($map,52.297275,4.8616077,52.28728,4.85771));
#print_path($map,Astar($map,52.297275,4.8616077,52.28728,4.85771,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.28728,4.85771,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.28728,4.85771,'car'));

#print_path($map,Astar($map,52.297275,4.8616077,52.311311,4.8468799,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.4176385,4.8708703,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.278473,4.846854,'car'));
#print_path($map,Astar($map,52.297275,4.8616077,52.4176385,4.8708703,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2933,4.8588));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'foot'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'bicycle'));
#print_path($map,Astar($map,52.297275,4.8616077,52.2935821,4.8593675,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2932895,4.8544163));
#print_path($map,Astar($map,52.297275,4.8616077,52.2932895,4.8544163));
#print_path($map,Astar($map,52.297275,4.8616077,52.2871944,4.8503294,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.2871944,4.8503294,'car'));

###Coentunnel
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'foot'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.4184,4.8724,'bicycle'));
#print_path($map,Astar($map,52.4184,4.8724,52.2973969,4.8620826,'car'));
#print_path($map,Astar($map,52.4184,4.8724,52.2973969,4.8620826,'bicycle'));

## Croon Delft
print_path($map,Astar($map,52.2973969,4.8620826,52.27909,4.86404,'car'));
print_path($map,Astar($map,52.27909,4.86404,51.9972199,4.3855367,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367,'car'));
#print_path($map,Astar($map,51.9972199,4.3855367,52.2973969,4.8620826,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367,'bicycle'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.9972199,4.3855367));
#print_path($map,Astar($map,52.4184,4.8724,51.9972199,4.3855367,'car'));

## ProRail
#print_path($map,Astar($map,52.2973969,4.8620826,52.087473,5.115715,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,52.087473,5.115715,'bicycle'));
#print_path($map,Astar($map,52.087473,5.115715,52.2973969,4.8620826,'car'));

## ICT Barendrecht
#print_path($map,Astar($map,52.2973969,4.8620826,51.8503978,4.5091717,'car'));
#print_path($map,Astar($map,51.8503978,4.5091717,52.2973969,4.8620826,'car'));
#print_path($map,Astar($map,52.2973969,4.8620826,51.8503978,4.5091717,'bicycle'));

## Roquebrune
#print_path($map,Astar($map,52.2973969,4.8620826,43.4046930,6.6792379,'car'));

###Brussel
#print_path($map,Astar($map,52.2973969,4.8620826,50.8417207,4.3832422,'car'));
}
$map->postprocess();

File: trunk/osm/astar.cpp
#include "osm.h"
#include <map>
#include <cstdlib>
#include <stdbool.h>
#include "grootcirkel.h"

#define EXTRACOST_FACTOR (100.0)

namespace osm {
  using namespace std;
  
  void Map::initRoute(const string &vehicle) {
    if (vehicle != "" && _profiles.find(vehicle) == _profiles.end())
        throw domain_error("Onbekend voertuig : "+vehicle);
    _vehicle = vehicle;
  }

  bool Map::wrong_direction(Node &nodex,Node &nodey,Way &ww,string onew) const {

    int dir = _con->getDirection(nodex.id(),nodey.id(),ww.id());
    if (dir == 0)
      throw runtime_error("geen directe verbinding tussen nodes");

    if (onew=="rev")
      return (dir == -1);
    else
      return (dir == 1);
  }
  
  long Map::curvecost(const long x,const long y,const long p) {
    if (p == 0)
      return 0;
    Node &ndx=nodes(x);
    Node &ndy=nodes(y);
    Node &ndp=nodes(p);
    
    double dx1 = ndy.lon()-ndx.lon();
    double dy1 = ndy.lat()-ndx.lat();
    double dx2 = ndx.lon()-ndp.lon();
    double dy2 = ndx.lat()-ndp.lat();
    
    double h1 = 180.0 * atan2(dy1,dx1) / PI;
    double h2 = 180.0 * atan2(dy2,dx2) / PI;
    double dh = fabs(h2-h1);
    if (dh > 180) 
      dh = 360 - dh;

    if (dh < 90)
      return 3 * (dh/90);
    else if (dh < 120) 
      return 3 + 4*(dh-90)/30;
    else if (dh < 150)
      return 5 + 5*(dh-120)/30;

    return 10 + 10 * (dh-150)/30;
  }
  
  double Map::cost(const long x,const long y,const long prevnode,const bool ignoreExtra) { 
    _costcounter++;
    double dist = distance(x,y);
    if (_vehicle == "") return dist;

    long extracost = 0; 
    double speed = _profiles[_vehicle].maxspeed();
    long w = getConnectingWay(x,y);
    if (w == 0)
      throw range_error("Kan weg niet vinden");
    osm::Way &ww = ways(w);
    string hw;
    
    long cnt = ww.getNodesCount() - 1;
    if (cnt < 1) cnt = 1;
    
// type weg

    try {
      hw = ww["highway"];
    } catch (range_error &ex) {
      try {
	if (ww["route"] == "ferry") {
	  hw = "unclassified";
	  speed = 12;    // vehicle speed is niet belangrijk
 	  extracost= 600 / cnt; // 10 minuten vertraging
	}
      } catch (range_error &ex) {
	return INFINITY;
      }
    }

// bepaal maximum snelheid
// eerst uit de weg halen anders de standaard snelheid voor zo'n weg
    
    try {
      double maxspeed = atol(ww["maxspeed"].c_str());
      if (maxspeed < speed)
	speed = maxspeed;
    } catch (range_error &ex) {
      double defspeed = _highways[hw].speed();
      if (defspeed < speed)
	speed = defspeed;
    }

    if (speed == 0)
      return INFINITY;

// basis kost is de tijd die je er op nominale snelheid over doet

    double kost = dist *3.6 / speed;

// beperkingen aflopen

    string access;
    string oneway;
    try { access = ww["access"];} catch (range_error &ex) {access="yes";}
    try { oneway = ww["oneway"];} catch (range_error &ex) {oneway="";}
    try {
      if (ww["junction"] == "roundabout" && oneway == "")
	oneway = "yes";
    } catch (range_error &ex) {}

    extracost += _highways[hw].extracost()*dist/EXTRACOST_FACTOR;   // extracost gaat per kilometer als ze op een weg slaan    
    Node &nodey = nodes(y);                               // node waar we naar toe gaan

    if (_vehicle == "foot") {
      string fa;
      try { fa = ww["foot"];} catch (range_error &ex) {fa="";}
      if (fa == "") {
	try {
	  if (ww["motorroad"] == "yes") {
	    fa="no";
	  }
	} catch (range_error &ex) {}	
      }
      if ((fa == "no") || (access == "no" && fa != "yes"))
	return INFINITY;
      try { extracost += _profiles[_vehicle].allowed(hw)*dist/EXTRACOST_FACTOR;} catch (range_error &ex) {return INFINITY;}
      try { extracost += _highways[nodey["highway"]].extracost();} catch (range_error &ex) {};
    } else if (_vehicle == "bicycle") {
      string cw;
      try { cw = ww["cycleway"];} catch (range_error &ex) {cw="";}
      if (cw == "opposite_lane" || cw == "opposite_track" || cw == "both")
	cw = "opposite";
      string ca;
      try { ca = ww["bicycle"];} catch (range_error &ex) {ca="";}
      if (ca == "") {
	try {
	  if (ww["motorroad"] == "yes") {
	    ca="no";
	  }
	} catch (range_error &ex) {}	
      }

      if (ca != "yes") {  // als expliciet toegestaan dan geen extracost meenemen (dus extracost voor dit type weg is dan 0 ongeacht profiel waarde)
	try { 
	  extracost += _profiles[_vehicle].allowed(hw)*dist/EXTRACOST_FACTOR;
	} catch (range_error &ex) {
	  if (cw == "") 
	    return INFINITY;
	}
      }
      if (cw != "")
	extracost = 0; // prefer cycle track or lane
      else {
        if ((ca == "no") || (access == "no" && ca != "yes"))
	  return INFINITY;
      }
      if (oneway != "" && cw != "opposite") {
	if (wrong_direction(nodes(x),nodey,ww,oneway))
	  return INFINITY;
     }
      try {extracost += _highways[nodey["highway"]].extracost();} catch (range_error &ex) {};
    } else if (_vehicle == "car") {
      string ma;
      try { ma = ww["motorcar"];} catch (range_error &ex) {ma="";}
      if ((ma == "no") || (access == "no" && ma != "yes"))
	return INFINITY;
      try { extracost += _profiles[_vehicle].allowed(hw)*dist/EXTRACOST_FACTOR;} catch (range_error &ex) {return INFINITY;}
      if (oneway != "") {
	if (wrong_direction(nodes(x),nodey,ww,oneway))
	  return INFINITY;
      }
      try { extracost += _highways[nodey["highway"]].extracost();} catch (range_error &ex) {};
    }
    
//  als nog steeds access = no dan mag het echt niet

    if (access == "no")
      return INFINITY;

    extracost += curvecost(x,y,prevnode) * _profiles[_vehicle].curvefactor();
// nog wat beperkingen  (als waarde * dan voor alle niet expliciet gespecificeerde in profiel)
      
    try { 
      extracost += _profiles[_vehicle].traffic_calming(nodey["traffic_calming"]);
    } catch (range_error &ex) {
      try { extracost += _profiles[_vehicle].traffic_calming("*");} catch (range_error &ex1) {};
    };
    try { 
      extracost += _profiles[_vehicle].barrier(nodey["barrier"]);
    } catch (range_error &ex) {
      try { extracost += _profiles[_vehicle].barrier("*");} catch (range_error &ex1) {};
    };
    try { 
      extracost += _profiles[_vehicle].barrier(nodey["highway"]);
    } catch (range_error &ex) {
      try { extracost += _profiles[_vehicle].highway("*");} catch (range_error &ex1) {};
    };
    
    if (extracost >= INFINITY)
      return INFINITY;

    if (!ignoreExtra)
      kost += extracost;
    if (kost > INFINITY)
      kost = INFINITY;
    
    return kost;
  }
  
  double Map::calc_h_score(const long n1,const long n2) {
    double dist = distance(n1,n2);
    if (_vehicle == "") return dist;

    return dist * 1.75 * 3.6 / _profiles[_vehicle].avgspeed();
  }

  long bestpoints[3];
  double initialdistance;
  double maxperc = 0;
  
  long Map::AstarHelper(int set,long goal,set_type &openset,set_type &closedset,score_type &f,score_type &g,score_type &h,score_type &d,route_type &to,const bool ignoreExtra) {
    long xs = 0;
    set_type::iterator k;
    
    for (k=openset.begin();k!=openset.end();k++) {
      if (xs == 0) 
	xs = k->first;
      if (f[k->first] < f[xs])
	xs = k->first;
    }
    if (xs == 0)  // no more nodes to process should not happen in our situation 
      throw runtime_error("xs = 0");
      
//    cerr << "xs = " << xs << endl;
    
    bestpoints[set] = xs;
    double newdistance = distance(bestpoints[1],bestpoints[2]);
    double curperc = 100.0*(initialdistance - newdistance)/initialdistance;
    maxperc = max(maxperc,curperc);

    k = closedset.find(xs);
    if (k != closedset.end() && k->second != set) {
      cout << "S set = " << set << " xs match = " << xs << endl;
      return xs;
    }
    openset.erase(xs);
    closedset[xs]=set;

    long prevnode = 0;
    route_type::iterator pn;
    pn = to.find(xs);
    if (pn != to.end())
      prevnode = pn->second;

    vector<long> neighbours;
    getNeighbours(xs,neighbours);

    for (vector<long>::iterator yi=neighbours.begin();yi != neighbours.end();yi++) {
      long y = *yi;
      
//      cerr << "y = " << y << endl;

      if (closedset.find(y) == closedset.end()) { // && g[xs] != INFINITY) {
//        cerr << "not closed"<<endl;
	double tentative_g_score = g[xs] + (set==1?cost(xs,y,prevnode,ignoreExtra):cost(y,xs,prevnode,ignoreExtra));
	if (tentative_g_score >INFINITY)
	  tentative_g_score = INFINITY;
	bool tentative_is_better = false;

	if (openset.find(y) == openset.end()) {
	  openset[y] = 1;
//          cerr << "added to openset" << endl;
        }
        if (g.find(y) != g.end()) {
          if (tentative_g_score < g[y] || g[y] == 0) {
            tentative_is_better = true;
	        cout << "verbetering van " << y << endl << "oud = " << g[y] << " nieuw = " << tentative_g_score << endl;
//
// alle nodes die vanaf y bereikbaar zijn moeten opnieuw worden berekend (behalve xs1)
// dus als ze al in closedset staan daar weer uit verwijderen
//	  
            vector<long> ynb;
            getNeighbours(y,ynb);
            for (vector<long>::iterator y1i=ynb.begin();y1i != ynb.end();y1i++) {
              long y1=*y1i;
              if (y1 != xs) {
                k = closedset.find(y1);
                if (k != closedset.end()) {
                  if (k->second == set) {
                    cout << "   opnieuw open " << y1 << endl;
//                    closedset.erase(y1);
                  }
                }
              }
            }
          }
        } else { 
          tentative_is_better = true;
//          cerr << "bestond nog niet" << endl;
	}
        
	  string name;
	  string ref;
	  long w = getConnectingWay(xs,y);
	  osm::Way &ww = ways(w);
	  try { name = ww["name"];} catch (range_error &ex) {name="";}
	  try { ref = ww["ref"];} catch (range_error &ex) {ref="";}
	if (tentative_is_better) {
	  to[y] = xs;
	  g[y] = tentative_g_score;
	  h[y] = calc_h_score(y,goal);
	  d[y] = distance(y,goal);
	  f[y] = g[y] + h[y];
	  if (f[y] > INFINITY)
	    f[y] = INFINITY;
	  string name;
	  string ref;
	  long w = getConnectingWay(xs,y);
	  osm::Way &ww = ways(w);
	  try { name = ww["name"];} catch (range_error &ex) {name="";}
	  try { ref = ww["ref"];} catch (range_error &ex) {ref="";}
	  cout << set << " " << maxperc << " " <<  curperc ;
	  if (set == 1) 
	    cout << " " << xs << " " << y << " " << f[xs] << " " << g[y] << " " << h[y] << " " << f[y] << " "+name+" "+ref <<endl;
	  else
	    cout << " " << y << " " << xs << " " << f[xs] << " " << g[y] << " " << h[y] << " " << f[y] << " "+name+" "+ref <<endl;
	} else {
	  cout << set << " XXXXXX";
	  if (set == 1) 
	    cout << " " << xs << " " << y << " " << f[xs] << " " << g[y] << "/" << tentative_g_score << " " << h[y] << " " << f[y] << " "+name+" "+ref <<endl;
	  else
	    cout << " " << y << " " << xs << " " << f[xs] << " " << g[y] << "/" << tentative_g_score << " " << h[y] << " " << f[y] << " "+name+" "+ref <<endl;
        }
      }
    }
    return 0;
  }
  
  double Map::Astar(const long n1,const long n2,const string &vehicle,list<long> &route,const bool ignoreExtra) {
    set_type startset,goalset,closedset;
    score_type gs_score,hs_score,fs_score,ds_score, gg_score,hg_score,fg_score,dg_score;
    route_type came_from,goes_to;
    long xs = 0;

    initRoute(vehicle);
    startset[n1] = 1;
    goalset[n2] = 1;
    gs_score[n1] = 0;
    initialdistance = dg_score[n2] = ds_score[n1] = distance(n1,n2);
    bestpoints[1] = n1;
    bestpoints[2] = n2;
    
    while (xs == 0 && (!startset.empty() || !goalset.empty())) {
      xs = AstarHelper(1,n2,startset,closedset,fs_score,gs_score,hs_score,ds_score,came_from,ignoreExtra);
      if (xs == 0)
        xs = AstarHelper(2,n1,goalset,closedset,fg_score,gg_score,hg_score,dg_score,goes_to,ignoreExtra);
    }

    route.clear();
    if (xs == 0)
        throw range_error("Niet mogelijk om een route te berekenen");
    route.push_back(xs);
    route_type::iterator rp;

    long x = xs;
    while ((rp = came_from.find(x)) != came_from.end()) {
      x = rp->second;
      route.push_front(x);
    }
    x = xs;
    while ((rp = goes_to.find(x)) != goes_to.end()) {
      x = rp->second;
      route.push_back(x);
    }

    return 0;
  }
  
  double Map::Astar(const long n1,const double lat2,const double lon2,const string &vehicle,list<long> &route,const bool ignoreExtra) {
    return 0;
  }
  
  double Map::Astar(const double lat1,const double lon1,const long n2,const string &vehicle,list<long> &route,const bool ignoreExtra) {
    return 0;
  }
  
  double Map::Astar(const double lat1,const double lon1,const double lat2,const double lon2,const string &vehicle,list<long> &route,const bool ignoreExtra) {
    return 0;
  }
}
